frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts

```diff
@@ -73,6 +82,46 @@ const renderLabels = (
     .html(labels.downstream);
 };
 
+/**
+ * The rendering approach we currently use is bidirectional:
+ * - Two graphs are rendered across the Y axis from a single root.
+ *   - For downstream, this flows as is from the API response
+ *   - For upstream however, we need to inverse the parent/child
+ *   relationships.
+ */
+export const reflowLineage = (items: LineageItem[]): LineageItem[] => {
+  const rootNode = items.find((n) => n.level === 0);
+
+  const lineageByKey = items.reduce(
+    (acc, i) => ({
+      ...acc,
+      [i.key]: i,
+    }),
+    {}
+  );
+
+  return items.reduce((acc, item) => {
+    if (item.level === 0) {
+      // Add the root node
+      acc.push(item);
```

Comment:
In general, I would suggest using ... instead of push most of the times. This might be problematic on some cases, but as a general rule, not mutating objects is a good rule of thumb.

Reply:
Hey, fully agree. Unfortunately there are a fair bit of in place mutations in this file, but very early I realized that d3 keeps internal references and actually copying/spreading an object quickly makes the behaviour unexpected ;o(
For this function though, it was completely ok to rework without mutations and have done so.